apps/prairielearn/elements/pl-string-input/pl-string-input.py

```diff
@@ -168,56 +160,42 @@ def render(element_html, data):
             html_params["missing_input"] = True
             html_params["parse_error"] = None
         else:
-            raw_submitted_answer = data["raw_submitted_answers"].get(name, None)
-            if raw_submitted_answer is not None:
-                html_params["raw_submitted_answer"] = pl.escape_unicode_string(
-                    raw_submitted_answer
-                )
+            html_params["raw_submitted_answer"] = raw_submitted_answer
+
+        score = data["partial_scores"].get(name, {"score": None}).get("score", None)
 
-        partial_score = data["partial_scores"].get(name, {"score": None})
-        score = partial_score.get("score", None)
         if score is not None:
-            try:
-                score = float(score)
-                if score >= 1:
-                    html_params["correct"] = True
-                elif score > 0:
-                    html_params["partial"] = math.floor(score * 100)
-                else:
-                    html_params["incorrect"] = True
-            except Exception:
-                raise ValueError("invalid score" + score)
+            score_type, score_value = pl.determine_score_params(score)
+            html_params[score_type] = score_value
 
         html_params["error"] = html_params["parse_error"] or html_params.get(
             "missing_input", False
         )
 
-        with open("pl-string-input.mustache", "r", encoding="utf-8") as f:
-            html = chevron.render(f, html_params).strip()
+        return chevron.render(template, html_params).strip()
+
     elif data["panel"] == "answer":
         a_tru = pl.from_json(data["correct_answers"].get(name, None))
-        if a_tru is not None:
-            html_params = {
-                "answer": True,
-                "label": label,
-                "a_tru": a_tru,
-                "suffix": suffix,
-            }
-            with open("pl-string-input.mustache", "r", encoding="utf-8") as f:
-                html = chevron.render(f, html_params).strip()
-        else:
-            html = ""
-    else:
-        raise Exception("Invalid panel type: %s" % data["panel"])
+        if a_tru is None:
+            return ""
 
-    return html
+        html_params = {
+            "answer": True,
+            "label": label,
+            "a_tru": a_tru,
+            "suffix": suffix,
+        }
+
+        return chevron.render(template, html_params).strip()
 
+    assert_never(data["panel"])
 
-def parse(element_html, data):
+
+def parse(element_html: str, data: pl.QuestionData) -> None:
     element = lxml.html.fragment_fromstring(element_html)
     name = pl.get_string_attrib(element, "answers-name")
     # Get allow-blank option
-    allow_blank = pl.get_string_attrib(element, "allow-blank", ALLOW_BLANK_DEFAULT)
+    allow_blank = pl.get_boolean_attrib(element, "allow-blank", ALLOW_BLANK_DEFAULT)
```

Comment:
I think so yeah. I didn't double check, but before this change if you set `allow-blank="false"`, it will behave as if it is true because all nonempty strings in Python are "truthy" 💀 I think there was another attribute in this file with this issue.